shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -134,26 +180,34 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
    * @return The data converted into a format more suitable for caching.
    */
   override def convertColumnarBatchToCachedBatch(input: RDD[ColumnarBatch],
-                                       schema: Seq[Attribute],
-                                       storageLevel: StorageLevel,
-                                       conf: SQLConf): RDD[CachedBatch] = {
-    def putOnGpuIfNeeded(batch: ColumnarBatch): ColumnarBatch = {
-      if (batch.numCols() > 0 && !batch.column(0).isInstanceOf[GpuColumnVector]) {
-        val s = StructType(
-          schema.map(a => StructField(a.name, a.dataType, a.nullable, a.metadata)))
-        val gpuCB = new GpuColumnarBatchBuilder(s, batch.numRows(), batch).build(batch.numRows())
-        batch.close()
-        gpuCB
-      } else {
-        batch
+     schema: Seq[Attribute],
+     storageLevel: StorageLevel,
+     conf: SQLConf): RDD[CachedBatch] = {
+    val rapidsConf = new RapidsConf(conf)
+    if (rapidsConf.isSqlEnabled) {
+      def putOnGpuIfNeeded(batch: ColumnarBatch): ColumnarBatch = {
+        if (batch.numCols() > 0 && !batch.column(0).isInstanceOf[GpuColumnVector]) {
+          val s: StructType = getSchemaFromAttributeSeq(schema)
+          val gpuCB = new GpuColumnarBatchBuilder(s, batch.numRows(), batch).build(batch.numRows())
+          batch.close()
+          gpuCB
+        } else {
+          batch
+        }
       }
+
+      input.map(batch => {
+        withResource(putOnGpuIfNeeded(batch)) { gpuCB =>
+          compressColumnarBatchWithParquet(gpuCB)
+        }
+      })
+    } else {
+      throw new UnsupportedOperationException("Can't convert ColumnarBatch to CachedBatch on CPU")
```

Comment:
Yes, you are absolutely right. I haven't gotten to that part yet which is why I am throwing an exception. Is there a better way that you would like me to handle this?

Reply:
Sometimes we throw UnsupportedOperationException (or even IllegalStateException) when we don't expect to get here.  In this case, we do expect this could happen, so it would be good to put a comment, probably a TODO, explaining that this needs to be done at some point, ideally pointing to the feature request tracking that it needs to be done.